### DISCUSSION TITLE
Linux Stongswan miss one policy?

### DISCUSSION THREAD

**Discussion by konachan700** [strongswan/strongswan#879] [Q&A]
**System (please complete the following information):**
 - OS: OpenSuSe 15.3
 - Kernel version (if applicable): 5.3.18-59.40-default
 - strongSwan version(s): 5.9.5
 - Tested/confirmed with the latest version: yes

**Describe the bug**
I created a IKEv2 server for access to server network 192.168.99.0/24. All devices (MacOS, iOS, Win*) connect normally. They have access to server network, and server can ping all devices. 
But linux devices with strongswan don't have access to server network. Server can ping linux devices, but devices can't ping server. If I do `iptables -t nat -A POSTROUTING -d 192.168.99.0/24 -j SNAT --to-source 192.168.99.121 (virtual iface address)`, client start to have access to server network. But It's not right way)

Client strongswan create policy only for src 192.168.99.N/32 (virtual iface) dst 192.168.99.0/24, and don't create policy for src 192.168.0.N/32 (default local iface ip) dst 192.168.99.0/24

**Expected behavior**
Linux system should to have access to server network

**Additional context**
[ip xfrm pol.txt](https://github.com/strongswan/strongswan/files/8001919/ip.xfrm.pol.txt)
[server-ipsec.conf.txt](https://github.com/strongswan/strongswan/files/8001928/server-ipsec.conf.txt)

**Comment by tobiasbrunner** [comment]
> and don't create policy for src 192.168.0.N/32 (default local iface ip) dst 192.168.99.0/24

Why would you expect that? The point is to use the virtual IP and not the physical one. Check if you have a route in table 220 `ip route list table 220` and read the log on the client.

**Reply by konachan700** [reply]
> Why would you expect that?

In other OS it works)

```
localhost:/home/user # ip route list table 220
throw 172.17.0.0/16 proto static 
throw 192.168.0.0/24 proto static 
throw 192.168.0.1 proto static 
throw 192.168.3.0/24 proto static 
192.168.99.0/24 via 192.168.0.1 dev wlan0 proto static src 192.168.99.204
```

ipsec don't catch traffic if source 192.168.0.124:
```
user@localhost:~> ping 192.168.99.101
PING 192.168.99.101 (192.168.99.101) 56(84) bytes of data.
^C
--- 192.168.99.101 ping statistics ---
3 packets transmitted, 0 received, 100% packet loss, time 2049ms
```
```
localhost:/home/user # tcpdump -ni wlan0 icmp
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on wlan0, link-type EN10MB (Ethernet), capture size 262144 bytes
13:43:58.342248 IP 192.168.0.124 > 192.168.99.101: ICMP echo request, id 83, seq 1, length 64
13:43:59.368152 IP 192.168.0.124 > 192.168.99.101: ICMP echo request, id 83, seq 2, length 64
13:44:00.392138 IP 192.168.0.124 > 192.168.99.101: ICMP echo request, id 83, seq 3, length 64
```

but catch traffic, if source 192.168.99.204
```
user@localhost:~> ping 192.168.99.101
PING 192.168.99.101 (192.168.99.101) 56(84) bytes of data.
64 bytes from 192.168.99.101: icmp_seq=1 ttl=127 time=35.9 ms
64 bytes from 192.168.99.101: icmp_seq=2 ttl=127 time=61.9 ms
64 bytes from 192.168.99.101: icmp_seq=3 ttl=127 time=32.9 ms
^C
--- 192.168.99.101 ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2002ms
rtt min/avg/max/mdev = 32.919/43.603/61.945/13.029 ms
```
```
localhost:/home/user # tcpdump -ni wlan0 icmp
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on wlan0, link-type EN10MB (Ethernet), capture size 262144 bytes
13:46:25.328483 IP 192.168.99.101 > 192.168.99.204: ICMP echo reply, id 84, seq 1, length 64
13:46:26.355559 IP 192.168.99.101 > 192.168.99.204: ICMP echo reply, id 84, seq 2, length 64
13:46:27.327634 IP 192.168.99.101 > 192.168.99.204: ICMP echo reply, id 84, seq 3, length 64

```

**Reply by tobiasbrunner** [reply]
> In other OS it works)

Probably not the way you think. None of the other platforms will use their physical IP.

> 192.168.99.0/24 via 192.168.0.1 dev wlan0 proto static src 192.168.99.204

This looks OK. It should cause `192.168.99.204` to get used as source when sending packets to IPs in `192.168.99.0/24`. Check with e.g. `ip route get 192.168.99.101`.

Unless there is another route that overrides the above or the routing rules are not correct. So if the above command does not return the correct IP, check the rules with `ip rule` and all routing tables with `ip route list table all`. Also make sure you don't have any NAT rules on your system already that force the physical IP for all traffic (i.e. check `iptables-save`).

> ipsec don't catch traffic if source 192.168.0.124:

It can't, because the IPsec policy does not include that IP (and it shouldn't, as traffic to that IP can't be routed back to your client properly, you may want to read [ForwardingAndSplitTunneling](https://wiki.strongswan.org/projects/strongswan/wiki/ForwardingAndSplitTunneling)).

**Reply by konachan700** [reply]
> Check with e.g. ip route get 192.168.99.101
```
localhost:/home/user # ip route get 192.168.99.101
192.168.99.101 via 192.168.0.1 dev wlan0 src 192.168.0.124 uid 0 
    cache 
```

> check the rules with ip rule

```
localhost:/home/user # ip rule
0:      from all lookup local
10:     from all fwmark 0xdc lookup 220
220:    from all lookup 220
32766:  from all lookup main
32767:  from all lookup default
```

> ip route list table all
```
localhost:/home/user # ip route list table all
throw 172.17.0.0/16 table 220 proto static 
throw 192.168.0.0/24 table 220 proto static 
throw 192.168.0.1 table 220 proto static 
throw 192.168.3.0/24 table 220 proto static 
192.168.99.0/24 via 192.168.0.1 dev wlan0 table 220 proto static src 192.168.99.204 
default via 192.168.0.1 dev wlan0 proto dhcp metric 600 
x.x.x.x via 192.168.0.1 dev wlan0 proto static metric 600 
172.17.0.0/16 dev docker0 proto kernel scope link src 172.17.0.1 
192.168.0.0/24 dev wlan0 proto kernel scope link src 192.168.0.124 metric 600 
192.168.0.1 dev wlan0 proto static scope link metric 600 
192.168.3.0/24 dev br0 proto kernel scope link src 192.168.3.2 metric 425 linkdown 
broadcast 127.0.0.0 dev lo table local proto kernel scope link src 127.0.0.1 
local 127.0.0.0/8 dev lo table local proto kernel scope host src 127.0.0.1 
local 127.0.0.1 dev lo table local proto kernel scope host src 127.0.0.1 
broadcast 127.255.255.255 dev lo table local proto kernel scope link src 127.0.0.1 
broadcast 172.17.0.0 dev docker0 table local proto kernel scope link src 172.17.0.1 
local 172.17.0.1 dev docker0 table local proto kernel scope host src 172.17.0.1 
broadcast 172.17.255.255 dev docker0 table local proto kernel scope link src 172.17.0.1 
broadcast 192.168.0.0 dev wlan0 table local proto kernel scope link src 192.168.0.124 
local 192.168.0.124 dev wlan0 table local proto kernel scope host src 192.168.0.124 
broadcast 192.168.0.255 dev wlan0 table local proto kernel scope link src 192.168.0.124 
broadcast 192.168.3.0 dev br0 table local proto kernel scope link src 192.168.3.2 linkdown 
local 192.168.3.2 dev br0 table local proto kernel scope host src 192.168.3.2 
broadcast 192.168.3.255 dev br0 table local proto kernel scope link src 192.168.3.2 linkdown 
local 192.168.99.204 dev wlan0 table local proto kernel scope host src 192.168.99.204 
local 192.168.99.204 dev tun0 table local proto kernel scope host src 192.168.99.204 
```

iptables is empy and allows all

**Reply by tobiasbrunner** [reply]
> 10:     from all fwmark 0xdc lookup 220

What's that about? And what's with the `tun0` device?

**Reply by konachan700** [reply]
> What's that about?
it's artifact of previous experiments)) 

tun0 it's a openvpn device, it's down, but strongswan install duplicate ip on it)
```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc pfifo_fast state DOWN group default qlen 1000
    link/ether f0:de:f1:9b:5e:a4 brd ff:ff:ff:ff:ff:ff
    altname enp3s0
3: wlan0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether f8:16:54:e5:96:b0 brd ff:ff:ff:ff:ff:ff
    altname wlp2s0
    inet 192.168.0.124/24 brd 192.168.0.255 scope global dynamic noprefixroute wlan0
       valid_lft 66749sec preferred_lft 66749sec
    inet 192.168.99.204/32 scope global wlan0
       valid_lft forever preferred_lft forever
4: br0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default qlen 1000
    link/ether b6:0a:8c:24:0e:54 brd ff:ff:ff:ff:ff:ff
    inet 192.168.3.2/24 brd 192.168.3.255 scope global noprefixroute br0
       valid_lft forever preferred_lft forever
18: tun0: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UNKNOWN group default qlen 500
    link/none 
    inet 192.168.99.204/32 scope global noprefixroute tun0
       valid_lft forever preferred_lft forever
```

**Reply by tobiasbrunner** [reply]
Why does it have the same virtual IP assigned?

**Reply by konachan700** [reply]
> Why does it have the same virtual IP assigned?

I don't know 


before ipsec connection:
```
localhost:/home/user # ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc pfifo_fast state DOWN group default qlen 1000
    link/ether f0:de:f1:9b:5e:a4 brd ff:ff:ff:ff:ff:ff
    altname enp3s0
3: wlan0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether f8:16:54:e5:96:b0 brd ff:ff:ff:ff:ff:ff
    altname wlp2s0
    inet 192.168.0.124/24 brd 192.168.0.255 scope global dynamic noprefixroute wlan0
       valid_lft 86183sec preferred_lft 86183sec
4: br0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default qlen 1000
    link/ether 8e:e5:fa:e8:2f:3d brd ff:ff:ff:ff:ff:ff
    inet 192.168.3.2/24 brd 192.168.3.255 scope global noprefixroute br0
       valid_lft forever preferred_lft forever
5: docker0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 02:42:3f:b3:7f:38 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.1/16 brd 172.17.255.255 scope global docker0
       valid_lft forever preferred_lft forever
7: vethc517a32@if6: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master docker0 state UP group default 
    link/ether b2:59:ea:78:93:47 brd ff:ff:ff:ff:ff:ff link-netnsid 0
```
after:

```
localhost:/home/user # ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc pfifo_fast state DOWN group default qlen 1000
    link/ether f0:de:f1:9b:5e:a4 brd ff:ff:ff:ff:ff:ff
    altname enp3s0
3: wlan0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether f8:16:54:e5:96:b0 brd ff:ff:ff:ff:ff:ff
    altname wlp2s0
    inet 192.168.0.124/24 brd 192.168.0.255 scope global dynamic noprefixroute wlan0
       valid_lft 86168sec preferred_lft 86168sec
    inet 192.168.99.204/32 scope global wlan0
       valid_lft forever preferred_lft forever
4: br0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default qlen 1000
    link/ether 8e:e5:fa:e8:2f:3d brd ff:ff:ff:ff:ff:ff
    inet 192.168.3.2/24 brd 192.168.3.255 scope global noprefixroute br0
       valid_lft forever preferred_lft forever
5: docker0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 02:42:3f:b3:7f:38 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.1/16 brd 172.17.255.255 scope global docker0
       valid_lft forever preferred_lft forever
7: vethc517a32@if6: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master docker0 state UP group default 
    link/ether b2:59:ea:78:93:47 brd ff:ff:ff:ff:ff:ff link-netnsid 0
8: tun0: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UNKNOWN group default qlen 500
    link/none 
    inet 192.168.99.204/32 scope global noprefixroute tun0
       valid_lft forever preferred_lft forever
```

**Reply by tobiasbrunner** [reply]
Are you using the NetworkManager interface? Also, please post the complete client log.

**Reply by konachan700** [reply]
Yes, I try to use NM now. But I tried standalone strongswan too with same result.

[nm-log-1.txt](https://github.com/strongswan/strongswan/files/8002969/nm-log-1.txt)

**Reply by tobiasbrunner** [reply]
OK, that at least explains the `tun0` device. But I can't reproduce the rest.

> iptables is empy and allows all

That might not be true if nftables is used (try `iptables-legacy-save` too).

**Reply by konachan700** [reply]
I rebooted pc and iptables look like this: 
[iptables-1.txt](https://github.com/strongswan/strongswan/files/8003114/iptables-1.txt)

I tried to `iptables -X && iptables -F nat && iptables -F mangle && iptables -F filter` but I don't see any effect for ipsec.

**Reply by tobiasbrunner** [reply]
I can't reproduce this here using the NM plugin. I've also tried with a similar routing rule with fwmark and also the throw routes. The latter are installed by the [_bypass-lan_ plugin](https://wiki.strongswan.org/projects/strongswan/wiki/Bypass-lan), by the way, just in case you want to try if disabling that for charon-nm has any effect. To do the latter add this to strongswan.conf:
```
charon-nm {
  plugins {
    bypass-lan {
      load = no
    }
  }
}
```

I think SUSE uses firewalld, not sure how exactly it installs netfilter rules, but maybe they don't show up in iptables. Perhaps check `firewall-cmd --list-all`.

**Reply by konachan700** [reply]
> check firewall-cmd --list-all

oh, you sent me to a good direction... I fogret about firewall-cmd. If we want to use ipsec in opensuse, we must allow it in this firewall for selected zone. For client too)) Client is prohibited by default.
P.S. Rules really wasn't showed in the iptables-save output. I used to think that firewalld use iptables, but now I know if it false...

Thank you.

**Reply by tobiasbrunner** [reply]
Interesting. I wonder how exactly this made the route lookup fail, though.

**Reply by konachan700** [reply]
I don't know how... Suse firewall by default doesn't block output esp traffic, doesn't block output udp, but blocks all ipsec inner traffic)) and brokes routes) 
I repeat this bug in the 3 PCs with opensuse, and i got identical result.

**Reply by tobiasbrunner** [reply]
So what exactly did you need to change in firewalld? Just enable the `ipsec` service? Or was there anything more specific?

**Reply by konachan700** [reply]
Yes, I allowed ipsec in firewalld.
I think, it's a firewalld bug. I have updated my system two hours ago and ipsec stop working again. Icmp worked, but tcp/udp in ipsec was blocked. I stopped firewalld and ipsec work perfectly now.
